### PR TITLE
Issue/31555 add status field coupons api

### PR DESCRIPTION
**Pull request by anitaa1990**
Addresses part 1 of #31555. 

## All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

## Background

The Woo mobile app is working on adding coupon management to the app. As part of that project, there are a few enhancements in the GET Coupon(s) API endpoint that we would need to make. 
- We would need to display the status of the coupon. It looks like the GET Coupons and GET Coupon endpoints don't provide the status of the field. But this is available in `wp-admin`.
- It looks like only published coupons are available in the API response. It would be nice to list all coupons available for the store, regardless of the status, since Core displays it that way. 
-  Update the status of a coupon from the API.

### Designs
Screenshot 2022-01-05 at 4 03 09 PM
Screenshot 2022-01-05 at 4 03 52 PM

## Proposed Changes
This PR addresses the first point from the above list. i.e. add the `status` field to the GET Coupon(s) API response. Using an example JSON, this is how it would look like:

```
{
        "id": 13,
        "code": "test",
        "amount": "100.00",
        "status": "publish",   <-- New property added here
        "date_created": "2022-01-04T07:11:29",
        "date_created_gmt": "2022-01-04T07:11:29",
        "date_modified": "2022-01-05T09:45:28",
        "date_modified_gmt": "2022-01-05T09:45:28",
        "discount_type": "fixed_cart",
       ...
}
```

This change would allow us to reliably show the status of the coupon (as displayed in the designs). The rest of the blockers mentioned above will be taken care of in a separate PR. 

## How to test the changes in this Pull Request:
This is just a suggestion but please feel free to test other scenarios. 🙂

1. Add a new coupon in wp-admin → Coupons → Add New.
2. Publish the coupon (since only published coupons will be available in the API response).
3. In the API, run `GET /wp-json/wc/v3/coupons`.
4. Confirm that the `status` field is available in the API response.

## Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

## Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

Adds `status` field to the `GET /coupons` endpoint. This will display the status of the coupon. 

## Questions
- I have marked this PR as draft for now since I'm not sure how to go about adding unit tests for this PR. I have updated some existing unit tests for coupons. But these tests seem to be part of the legacy package. Should I create a new class for `class-wc-coupon-data-store-cpt.php` or should I continue to use `plugins/woocommerce/tests/legacy/unit-tests/coupon/data-store.php`?
- Since this PR adds changes to the data in the coupons section, I feel like we should also update the related documentation [here](https://github.com/woocommerce/woocommerce/wiki/Coupon-Data). I can do that in this PR, if this is fine?

### CODE REVIEW

File: plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php [hunk 1]
```diff
@@ -123,6 +123,7 @@ public function read( &$coupon ) {
 			array(
 				'code'                        => $post_object->post_title,
 				'description'                 => $post_object->post_excerpt,
+				'status'	                  => $post_object->post_status,
```
Comment (joshbetz) [inline plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php:126]:
```suggestion
				'status'                 => $post_object->post_status,
```

Comment (joshbetz) [inline plugins/woocommerce/includes/data-stores/class-wc-coupon-data-store-cpt.php]: Minor: Fixes an indentation issue

```suggestion
				'status'	                      => $post_object->post_status,
```

File: plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php [hunk 2]
```diff
@@ -347,6 +348,11 @@ public function get_item_schema() {
 					'type'        => 'string',
 					'context'     => array( 'view', 'edit' ),
 				),
+				'status'                      => array(
```
Comment (joshbetz) [inline plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php]: Minor: Fixes an indentation issue
```suggestion
				'status' => array(
```

File: plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php [hunk 3]
```diff
@@ -347,6 +348,11 @@ public function get_item_schema() {
...
+					'description' => __( 'The status of the coupon. Should always be draft, published or pending review', 'woocommerce' ),
```
Comment (joshbetz) [inline plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php]: Minor: Oxford comma https://make.wordpress.org/docs/style-guide/punctuation/commas/#oxford-serial-commas
```suggestion
					'description' => __( 'The status of the coupon. Should always be draft, published, or pending review', 'woocommerce' ),
```

**Review by joshbetz — CHANGES_REQUESTED** [review]
A few minor style things, but otherwise this looks great to me.

**Comment by anitaa1990** [comment]
Thank you for the review @joshbetz 🙇‍♀️ I've updated the suggestions requested. Is there a command to run those indentation tests locally before I open the PR? I could do that in future PRs 😅

File: plugins/woocommerce/tests/legacy/unit-tests/coupon/data.php [hunk 4]
```diff
@@ -174,6 +174,7 @@ public function test_coupon_getters_and_setters() {
 			'description'                => 'hello world',
 			'discount_type'              => 'percent',
 			'amount'                     => 10.50,
+			'status'					 => 'publish',
```
Comment (joshbetz) [inline plugins/woocommerce/tests/legacy/unit-tests/coupon/data.php]:
```suggestion
			'status'                       => 'publish',
```

**Review by joshbetz — DISMISSED** [review]

**Comment by anitaa1990** [comment]
Thanks for approving the PR @joshbetz 😄 I'm not sure how to merge this! There seems like one of the checks was cancelled and till it is completed, I'm not able to merge the PR. Do you know what I need to do?

**Review by jeffstieler — APPROVED** [review]
👍

**Comment by DanielLanger** [comment]
@jeffstieler @anitaa1990 im not seeing this in the API docs https://woocommerce.github.io/woocommerce-rest-api-docs/#coupons

nor am I able to use the field. was this deployed?

**Pull request merged**